### PR TITLE
Add initial support for analyzing AMDGPU binaries

### PR DESCRIPTION
**Pull request by bbiiggppiigg**
This pull request contains the initial effort to support analyzing AMD GPU binaries.

Right now decoding is only supported for the kernels using the VEGA instruction set, and is not even complete because we yet to get the documents from AMD, and all the code is tested only with gfx-803.

When we want to add the support for another AMDGPU ISA, the structure of the code will likely undergo huge changes.

### CODE REVIEW

File: dataflowAPI/rose/semantics/Registers.C
Comment (hainest) [inline dataflowAPI/rose/semantics/Registers.C:651]: Leaky singletons are my least favorite singleton. Could we change this to

    RegisterDictionary const& RegisterDictionary::dictionary_amdgpu() {
        static std::once_flag initialized;
        static RegisterDictionary regs;

        std::call_once(...)

        return regs;
    }

or

    RegisterDictionary const* RegisterDictionary::dictionary_amdgpu() {
        static std::once_flag initialized;
        static RegisterDictionary regs;

        std::call_once(...)

        return &regs;
    }

if you need to preserve the pointer semantics of the return value? Both versions are safe in terms of lifetimes and threading.

File: dataflowAPI/src/SymbolicExpansion.C
Comment (hainest) [inline dataflowAPI/src/SymbolicExpansion.C:123]: Do we really want to just swallow this exception?

File: instructionAPI/src/ArchSpecificFormatters.C
Comment (hainest) [inline instructionAPI/src/ArchSpecificFormatters.C:164]: As of C++11, having the destination iterator be part of the input range is undefined behavior. Changing this to a simple range-based loop is the easy fix.

    for(auto &c : ret) ::toupper(c);

Also, there's no need to make a copy of the parameter since it's passed by value.

File: instructionAPI/src/Ternary.C
Comment (hainest) [inline instructionAPI/src/Ternary.C:97]: Same as above.

File: symtabAPI/src/Object-elf.C
Comment (hainest) [inline symtabAPI/src/Object-elf.C:682]: Empty block. Can be removed?

File: symtabAPI/src/debug.C
Comment (hainest) [inline symtabAPI/src/debug.C:38]: Let's not enable this by default. It's quite noisy.

File: instructionAPI/src/Ternary.C
Comment (hainest) [inline instructionAPI/src/Ternary.C:117]: Unreachable code.

File: parseAPI/src/IA_amdgpu.C
Comment (hainest) [inline parseAPI/src/IA_amdgpu.C:139]: If this code isn't needed, then let's remove it.

File: dataflowAPI/rose/SgAsmAmdgpuInstruction.h
Comment (hainest) [inline dataflowAPI/rose/SgAsmAmdgpuInstruction.h:29]: Only the first `public` is needed.

**Review by hainest — CHANGES_REQUESTED** [review]
There is lots of commented-out code. If it's not needed, then let's go ahead and remove it. If it's useful, then let's add a comment describing it.

A couple of notes on adding new classes:

1. If a class is not intended to be inherited from, then let's mark it `final` and remove the virtual specifiers from the member functions.

2. Base members that are overridden should be marked `override` to help the compiler enforce interface checking.

3. Thanks for using Doxygen comments!

File: dataflowAPI/rose/semantics/Registers.C
Comment (bbiiggppiigg) [reply dataflowAPI/rose/semantics/Registers.C:651]: This part I'm just following what was done with armv8.
I can probably fix the code to use the second pattern.
Can you explain to me what is problem with this code ? 
Like I'm not that familiar with C++, so  what would be the difference between the original version and the version you've suggested.

File: dataflowAPI/src/SymbolicExpansion.C
Comment (bbiiggppiigg) [reply dataflowAPI/src/SymbolicExpansion.C:123]: This is also what have been done for the rest of the architectures.(PPC64, AArch64,X86_64) and so on, where the fprintfs are commented out.
I guess the proper way would be to use some sort of debug printf instead of directly writing to stderr ?

File: instructionAPI/src/Ternary.C
Comment (bbiiggppiigg) [reply instructionAPI/src/Ternary.C]: This part I marked it as TODO as I'm not sure how should we bind an Ternary AST yet.

File: parseAPI/src/IA_amdgpu.C
Comment (bbiiggppiigg) [reply parseAPI/src/IA_amdgpu.C]: removed

File: dataflowAPI/rose/SgAsmAmdgpuInstruction.h
Comment (bbiiggppiigg) [reply dataflowAPI/rose/SgAsmAmdgpuInstruction.h]: Duplicates removed.

File: symtabAPI/src/debug.C
Comment (bbiiggppiigg) [reply symtabAPI/src/debug.C]: Restore to default = 0.

File: symtabAPI/src/Object-elf.C
Comment (bbiiggppiigg) [reply symtabAPI/src/Object-elf.C]: Empty block removed.

File: instructionAPI/src/ArchSpecificFormatters.C
Comment (bbiiggppiigg) [reply instructionAPI/src/ArchSpecificFormatters.C]: Fixed

File: instructionAPI/src/Ternary.C
Comment (bbiiggppiigg) [reply instructionAPI/src/Ternary.C]: Fixed

File: common/h/dyn_regs.h
Comment (bbiiggppiigg) [reply common/h/dyn_regs.h]: Fixed

File: dataflowAPI/src/SymbolicExpansion.C
Comment (mxz297) [reply dataflowAPI/src/SymbolicExpansion.C:123]: For this one, Dyninst should just swallow this exception. rose has more restrict requirements regarding instruction opcode, operand types, orders, and so on, and rose may also throw exception for unimplemented instructions. For all such error cases, this code will return an null AST ptr and a false return value. So, the user of Dyninst can detect this error and do corresponding error handling.

File: common/src/dyn_regs.C
Comment (mxz297) [inline common/src/dyn_regs.C:38]: @bbiiggppiigg Do we really need to use `<boost/format.hpp>`? @hainest Is there a std c++ equivalent?

**Comment by mxz297** [comment]
@bbiiggppiigg I find that there is some code about the instruction semantics. I did not know that you have started this work. So, what's the status of this work for instruction semantics? Specifically, can we get symbolic expression through dataflowAPI for certain AMD GPU instructions?

**Comment by bbiiggppiigg** [comment]
That is part of my effort trying to reason about function calls using indirect branching working, but it is still work in progress.
My target is to make the backward-slicing thing works. 

I primarily focused on the following instructions
s_setpc_b64, s_swappc_b64 : the instructions that updates the PC
s_getpc_b64 which stores the pc into some scalar register
s_add_u32/ s_addc_u32 : the instructions that are used to calculate the branch target

I'm not sure what do you expect to get using dataflowAPI, like can you give me an example of what you would like to see ?

**Comment by mxz297** [comment]
DataflowAPI has its interface for converting instructions to SSA-like assignments, slicing, and converting assignment to symbolic expressions. Resolving indirect control flow using all of mentioned interface. But, each of them can be used independently and can be used for other things.

**Comment by mxz297** [comment]
@bbiiggppiigg My goal was to only add preliminary support for AMD GPU analysis in terms of decoding instructions and parsing direct control flow. We have tested it with several AMD GPU binaries.

I would recommend that you split your work on dataflowAPI from the other components. Given that dataflowAPI support for AMD GPU has not been shown to work on any AMD GPU binary, I don't think we want to include that in a release.

File: dataflowAPI/rose/semantics/Registers.C
Comment (hainest) [reply dataflowAPI/rose/semantics/Registers.C:651]: I'm definitely not faulting you for using the existing code as a template here, even though the method is sub-optimal. I just want to make sure we are improving the code as we update it. The leaky single pattern (i.e., never freeing the instance) is only useful when there is a problem in ordering of destruction of static objects. That's not an issue here, so we can use better patterns that don't cause fault positives when run under a memory analysis tool (e.g., valgrind).

From the C++ side of things, C++11 added thread safety to function-local static variables. This means that `regs` is only constructed once (when the function is first called) and that initialization is thread-safe. This wasn't the case when the previous code was written (i.e., C++98). This new guarantee allows us to use a local variable with static storage duration instead of storing it on the heap. The call to `call_once` is still necessary because we are doing some non-trivial work that should only be done once. Does that help at all? It's one of the more technical changes made to the language, so I'm not surprised you are unfamiliar with it.

File: dataflowAPI/src/SymbolicExpansion.C
Comment (hainest) [reply dataflowAPI/src/SymbolicExpansion.C:123]: @mxz297 Sounds good. I just wanted to make sure this wasn't an accident.

File: common/src/dyn_regs.C
Comment (hainest) [reply common/src/dyn_regs.C:38]: I did a quick grep and couldn't find any usage of it. I think we can remove it.

> Is there a std c++ equivalent?

Yes, but it requires C++20.

**Comment by hainest** [comment]
I strongly suggest taking a look at the classes that were added here to improve type safety using the guidelines I mentioned (copied here for convenience).

A couple of notes on adding new classes:

1. If a class is not intended to be inherited from, then let's mark it final and remove the virtual specifiers from the member functions.

2. Base members that are overridden should be marked override to help the compiler enforce interface checking.

**Comment by bbiiggppiigg** [comment]
> @bbiiggppiigg My goal was to only add preliminary support for AMD GPU analysis in terms of decoding instructions and parsing direct control flow. We have tested it with several AMD GPU binaries.
> 
> I would recommend that you split your work on dataflowAPI from the other components. Given that dataflowAPI support for AMD GPU has not been shown to work on any AMD GPU binary, I don't think we want to include that in a release.

The problem here is that, even for running the funcDisassemble example alone, when an instruction is an indirect branch, it will trigger the IndirectAnalyzer in parseAPI, 

https://github.com/bbiiggppiigg/dyninst/blob/5c6f0bd2c3d3e4a26d702e7eaa566b1934487896/parseAPI/src/IndirectAnalyzer.C#L78-L83

which will in turn calls to the AssignmentConverter in dataflowAPI, and the IndirectAnalyzer  expect to see at least one assignment inserted into the assignment vector, so I see no clean way of separating out all dataflowAPI work.

**Comment by mxz297** [comment]
@bbiiggppiigg You can choose to not resolve indirect branch for AMD GPU binaries. The IndirectAnalyzer can simply directly return an empty set of edges for AMD GPU binaries.

File: dataflowAPI/rose/semantics/Registers.C
Comment (bbiiggppiigg) [reply dataflowAPI/rose/semantics/Registers.C]: So this part of the code is removed as an effort to separate out the work on dataflowAPI. I'll mark it as resolved here.

**Comment by bbiiggppiigg** [comment]
@mxz297 the changes in dataflowAPI have now been removed.

**Comment by hainest** [comment]
@bbiiggppiigg Is there anything remaining to be updated before testing?

**Comment by bbiiggppiigg** [comment]
I'm not aware of anything remaining to be updated.

**Comment by bbiiggppiigg** [comment]
@hainest In case that we're gonna support multiple AMDGPU architectures in the future, I'm considering whether I should rename the Arch_amdgpu to something like Arch_amdgpu_vega so I can later add something like Arch_amdgpu_rdna. Do you think it is necessary to change this before merging, or is it fine if these Arch specifiers are only used internally

**Comment by hainest** [comment]
This support is still very much experimental, so we make no guarantees about code stability. That said, it would probably be a good idea to split the architectures. We'll likely need to do that with CUDA in the future, as well. It would be nice to have a hierarchy of architectures (e.g., "amdgpu" would encompass vega, navi, etc.), but I don't think we can support that without major changes.

@mxz297 Any thoughts on this?

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR900

Looks good. We can merge now or wait until you have split the AMD GPU architectures.

**Comment by hainest** [comment]
ping.

**Comment by mxz297** [comment]
I think it is a good idea to rename `Arch_amdgpu` to `Arch_amdgpu_vega`. Otherwise, when we later add more ADM GPU architectures, we will break API by renaming the architecture enums.

**Comment by hainest** [comment]
Ok. I'll hold off on merging this until you've made those changes.

**Comment by bbiiggppiigg** [comment]
@hainest I've made changes according to the suggestions.

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR900&since_date=2020-11-24

Looks good.

**Pull request merged**
